src/Schedgen2/mpi_colls.py

```diff
@@ -56,103 +27,185 @@ def dissemination(comm_size, datasize, tag):
-        sources = sources[::-1]
-        destinations = sources
-        data_sizes_receive = data_sizes_receive[::-1]
-        data_sizes_send = data_sizes_send[::-1]
-        iterative_send_recv(
-            comm,
-            rank,
-            sources,
-            destinations,
-            data_sizes_receive,
-            data_sizes_send,
-            base_tag,
-            last_dependency=dependency,
-            compute_time_dependency=ctd,
-        )
 
+def outcast(
+    comm_size: int,
+    datasize: int,
+    tag: int = 42,
+    ptrn: str = "linear",
+    unbalanced: bool = False,
+    **kwargs,
+):
+    assert ptrn == "linear", "outcast only supports the linear communication pattern"
+    return linear(
+        comm_size=comm_size,
+        datasize=datasize,
+        tag=tag,
+        algorithm="outcast",
+        parallel=True,
+        randomized_data=unbalanced,
+        **kwargs,
+    )
 
-def ring_allreduce(comm, comm_size, datasize, base_tag, ctd=0):
-    for rank in range(0, comm_size):
-        chunk_size = (
-            datasize // comm_size
-            if datasize % comm_size == 0
-            else datasize // comm_size + 1
+
+def reduce(
+    comm_size: int,
+    datasize: int,
+    tag: int = 42,
+    ptrn: str = "binomialtree",
+    unbalanced: bool = False,
+    **kwargs,
+):
+    if ptrn == "binomialtree":
+        assert (
+            unbalanced == False
+        ), "binomialtree does not currently support randomized data"
+        return binomialtree(
+            comm_size=comm_size,
+            datasize=datasize,
+            tag=tag,
+            algorithm="reduce",
+            **kwargs,
         )
-        sources = [(rank - 1) % comm_size] * (comm_size - 1)
-        destinations = [(rank + 1) % comm_size] * (comm_size - 1)
-        data_sizes_receive = [chunk_size] * (comm_size - 1)
-        data_sizes_send = [chunk_size] * (comm_size - 1)
-        dependency = iterative_send_recv(
-            comm,
-            rank,
-            sources,
-            destinations,
-            data_sizes_receive,
-            data_sizes_send,
-            base_tag,
-            compute_time_dependency=ctd,
+    elif ptrn == "linear":
+        return linear(
+            comm_size=comm_size,
+            datasize=datasize,
+            tag=tag,
+            algorithm="reduce",
+            parallel=True,
+            randomized_data=unbalanced,
+            **kwargs,
         )
-        base_tag += 1
-        iterative_send_recv(
-            comm,
-            rank,
-            destinations,
-            sources,
-            data_sizes_send,
-            data_sizes_receive,
-            base_tag,
-            last_dependency=dependency,
-            compute_time_dependency=ctd,
+    else:
+        raise ValueError(f"reduce with pattern {ptrn} not implemented")
+
+
+def bcast(
+    comm_size: int,
+    datasize: int,
+    tag: int = 42,
+    ptrn: str = "binomialtree",
+    unbalanced: bool = False,
```

Comment:
MPI_Bcast has no argument for "unbalanced" anything, what is this supposed to do?

Reply:
I will rename the unbalanced to randomize_data, to make it more intuitive. Some benchmarks want to randomise the data sent to each process, which is enabled using that flag. This is not part of the MPI_Bcast, but I don't see an effective way to map this 1-to-1